File: java-exercise-fetcher/src/test/java/org/jesperancinha/twitter/client/FetcherThreadTest.java
package org.jesperancinha.twitter.client;

import com.twitter.hbc.core.event.Event;
import com.twitter.hbc.httpclient.BasicClient;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.TimeUnit;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.atLeastOnce;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.only;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
public class FetcherThreadTest {

    @Mock
    private BlockingQueue<String> queueMock;

    @Mock
    private ExecutorService executorServiceMock;

    @Mock
    private BasicClient clientMock;

    @Test
    public void testRun_whenFetchOk_thenReturnMessage() throws InterruptedException {
        MockitoAnnotations.initMocks(this);

        final List<String> allmessages = new ArrayList<>();

        when(queueMock.poll(1, TimeUnit.SECONDS)).thenReturn("I am a message!");

        final FetcherThread fetcherThread = FetcherThread.builder()
                .stringLinkedBlockingQueue(queueMock)
                .capacity(1)
                .allMessages(allmessages)
                .executorService(executorServiceMock)
                .client(clientMock)
                .build();

        fetcherThread.start();

        fetcherThread.join();

        assertThat(allmessages).isNotNull();
        assertThat(allmessages).hasSize(1);
        assertThat(allmessages).contains("I am a message!");

        verify(queueMock, only()).poll(1, TimeUnit.SECONDS);
        verify(clientMock, times(1)).connect();
        verify(clientMock, times(1)).stop();
        verify(clientMock, times(1)).isDone();
        verify(executorServiceMock, only()).shutdownNow();

    }

    @Test
    public void testRunWithCapacity5_whenFetchOk_thenReturnMessage() throws InterruptedException {
        MockitoAnnotations.initMocks(this);

        final List<String> allmessages = new ArrayList<>();

        when(queueMock.poll(1, TimeUnit.SECONDS)).thenReturn("I am a message!");

        final FetcherThread fetcherThread = FetcherThread.builder()
                .stringLinkedBlockingQueue(queueMock)
                .capacity(5)
                .allMessages(allmessages)
                .executorService(executorServiceMock)
                .client(clientMock)
                .build();

        fetcherThread.start();

        fetcherThread.join();

        assertThat(allmessages).isNotNull();
        assertThat(allmessages).hasSize(5);
        assertThat(allmessages).contains("I am a message!");

        verify(queueMock, times(5)).poll(1, TimeUnit.SECONDS);
        verify(clientMock, times(1)).connect();
        verify(clientMock, times(1)).stop();
        verify(clientMock, times(5)).isDone();
        verify(executorServiceMock, only()).shutdownNow();

    }


    @Test
    public void testRunWithCapacity5_whenFetchOkButOneNull_thenReturnStill5Messages() throws InterruptedException {
        MockitoAnnotations.initMocks(this);
        final String testMessage = "I am a message!";
        final List<String> allmessages = new ArrayList<>();

        final Stack<String> stringStack = new Stack<>();
        stringStack.push(testMessage);
        stringStack.push(testMessage);
        stringStack.push(testMessage);
        stringStack.push(testMessage);
        stringStack.push(testMessage);
        stringStack.push(null);
        when(queueMock.poll(1, TimeUnit.SECONDS)).thenAnswer(invocationOnMock -> stringStack.pop());

        final FetcherThread fetcherThread = FetcherThread.builder()
                .stringLinkedBlockingQueue(queueMock)
                .capacity(5)
                .allMessages(allmessages)
                .executorService(executorServiceMock)
                .client(clientMock)
                .build();

        fetcherThread.start();

        fetcherThread.join();

        assertThat(allmessages).isNotNull();
        assertThat(allmessages).hasSize(5);
        assertThat(allmessages).contains("I am a message!");
        assertThat(allmessages).containsOnly("I am a message!");

        verify(queueMock, times(6)).poll(1, TimeUnit.SECONDS);
        verify(clientMock, times(1)).connect();
        verify(clientMock, times(1)).stop();
        verify(clientMock, times(6)).isDone();
        verify(executorServiceMock, only()).shutdownNow();

    }

    @Test
    public void testRun_whenFetchMostlyEmptyAndCapacity_thenReturnOneMessage() throws InterruptedException {
        MockitoAnnotations.initMocks(this);

        final List<String> allmessages = new ArrayList<>();
        final String testMessage = "I am a message!";

        final Stack<String> stringStack = new Stack<>();
        stringStack.push(null);
        stringStack.push(null);
        stringStack.push(null);
        stringStack.push(null);
        stringStack.push(testMessage);
        stringStack.push(null);
        stringStack.push(null);
        stringStack.push(null);

        when(queueMock.poll(1, TimeUnit.SECONDS)).thenAnswer(invocationOnMock -> stringStack.pop());

        final FetcherThread fetcherThread = FetcherThread.builder()
                .stringLinkedBlockingQueue(queueMock)
                .capacity(1)
                .allMessages(allmessages)
                .executorService(executorServiceMock)
                .client(clientMock)
                .build();

        fetcherThread.start();

        fetcherThread.join();

        assertThat(allmessages).isNotNull();
        assertThat(allmessages).hasSize(1);
        assertThat(allmessages).contains("I am a message!");

        verify(queueMock, atLeastOnce()).poll(1, TimeUnit.SECONDS);
        verify(clientMock, times(1)).connect();
        verify(clientMock, times(1)).stop();
        verify(clientMock, atLeastOnce()).isDone();
        verify(executorServiceMock, only()).shutdownNow();

    }

    @Test
    public void testRun_whenClientIsDone_thenReturnNoMessage() throws InterruptedException {
        MockitoAnnotations.initMocks(this);

        final List<String> allmessages = new ArrayList<>();
        final Event event = mock(Event.class);
        when(clientMock.isDone()).thenReturn(true);
        when(clientMock.getExitEvent()).thenReturn(event);

        final FetcherThread fetcherThread = FetcherThread.builder()
                .stringLinkedBlockingQueue(queueMock)
                .capacity(1)
                .allMessages(allmessages)
                .executorService(executorServiceMock)
                .client(clientMock)
                .build();

        fetcherThread.start();

        fetcherThread.join();

        assertThat(allmessages).isNotNull();
        assertThat(allmessages).isEmpty();

        verify(queueMock, never()).poll(1, TimeUnit.SECONDS);
        verify(clientMock, times(1)).connect();
        verify(clientMock, times(1)).isDone();
        verify(clientMock, times(1)).getExitEvent();
        verify(clientMock, times(1)).stop();
        verify(executorServiceMock, only()).shutdownNow();

    }
}

File: twitter-fetcher/src/test/kotlin/org/jesperancinha/twitter/client/FetcherThreadTest.kt
package org.jesperancinha.twitter.client

import com.twitter.hbc.core.event.Event
import com.twitter.hbc.httpclient.BasicClient
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.nulls.shouldNotBeNull
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import java.util.*
import java.util.concurrent.BlockingQueue
import java.util.concurrent.ExecutorService
import java.util.concurrent.TimeUnit

@ExtendWith(MockKExtension::class)
class FetcherThreadTest {
    @MockK
    lateinit var queueMock: BlockingQueue<String>

    @MockK
    lateinit var executorServiceMock: ExecutorService

    @MockK(relaxed = true)
    lateinit var clientMock: BasicClient

    @Test
    fun testRun_whenFetchOk_thenReturnMessage() {
        val messages: Set<String> = HashSet()
        every { queueMock.poll(1, TimeUnit.SECONDS) } returns "I am a message!"
        val fetcherThread = FetcherThread.builder()
            .stringLinkedBlockingQueue(queueMock)
            .capacity(1)
            .allMessages(messages)
            .executorService(executorServiceMock)
            .client(clientMock)
            .build()
        fetcherThread.start()
        fetcherThread.join()
        Assertions.assertThat(messages).isNotNull
        Assertions.assertThat(messages).hasSize(1)
        Assertions.assertThat(messages).contains("I am a message!")
        verify { queueMock.poll(1, TimeUnit.SECONDS) }
        verify(exactly = 1) { clientMock.connect() }
        verify(exactly = 1) { clientMock.stop() }
        verify(exactly = 1) { clientMock.isDone }
        verify { executorServiceMock.shutdownNow() }
    }

    @Test
    fun testRunWithCapacity5_whenFetchOk_thenReturnMessage() {
        val allmessages: Set<String> = HashSet()
        every { queueMock.poll(1, TimeUnit.SECONDS) } returns "I am a message!"
        every { executorServiceMock.shutdownNow() } returns listOf()
        val fetcherThread = FetcherThread.builder()
            .stringLinkedBlockingQueue(queueMock)
            .capacity(5)
            .allMessages(allmessages)
            .executorService(executorServiceMock)
            .client(clientMock)
            .build()
        fetcherThread.start()
        fetcherThread.join()
        allmessages.shouldNotBeNull()
        Assertions.assertThat(allmessages).hasSize(1)
        Assertions.assertThat(allmessages).contains("I am a message!")
        verify(exactly = 5) { queueMock.poll(1, TimeUnit.SECONDS) }
        verify(exactly = 1) { clientMock.connect() }
        verify(exactly = 1) { clientMock.stop() }
        verify(exactly = 5) { clientMock.isDone }
        verify { executorServiceMock.shutdownNow() }
    }

    @Test
    fun testRunWithCapacity5_whenFetchOkButOneNull_thenReturnStill5Messages() {
        val testMessage = "I am a message!"
        val allmessages: Set<String> = HashSet()
        val stringStack = Stack<String?>()
        stringStack.push(null)
        stringStack.push(testMessage)
        stringStack.push(testMessage)
        stringStack.push(testMessage)
        stringStack.push(testMessage)
        stringStack.push(testMessage)
        stringStack.push(testMessage)
        every { queueMock.poll(1, TimeUnit.SECONDS) } returnsMany (stringStack)
        every { executorServiceMock.shutdownNow() } returns listOf()
        val fetcherThread = FetcherThread.builder()
            .stringLinkedBlockingQueue(queueMock)
            .capacity(5)
            .allMessages(allmessages)
            .executorService(executorServiceMock)
            .client(clientMock)
            .build()
        fetcherThread.start()
        fetcherThread.join()
        Assertions.assertThat(allmessages).isNotNull
        Assertions.assertThat(allmessages).hasSize(1)
        Assertions.assertThat(allmessages).contains("I am a message!")
        Assertions.assertThat(allmessages).containsOnly("I am a message!")
        verify(atLeast = 5, atMost = 6) { queueMock.poll(1, TimeUnit.SECONDS) }
        verify(exactly = 1) { clientMock.connect() }
        verify(exactly = 1) { clientMock.stop() }
        verify(atLeast = 5, atMost = 6) { clientMock.isDone }
        verify { executorServiceMock.shutdownNow() }
    }

    @Test
    fun testRun_whenFetchMostlyEmptyAndCapacity_thenReturnOneMessage() {
        val allmessages: Set<String> = HashSet()
        val testMessage = "I am a message!"
        val stringStack = Stack<String?>()
        stringStack.push(null)
        stringStack.push(null)
        stringStack.push(null)
        stringStack.push(null)
        stringStack.push(testMessage)
        stringStack.push(null)
        stringStack.push(null)
        stringStack.push(null)
        every { queueMock.poll(1, TimeUnit.SECONDS) }.returnsMany(stringStack)
        val fetcherThread = FetcherThread.builder()
            .stringLinkedBlockingQueue(queueMock)
            .capacity(1)
            .allMessages(allmessages)
            .executorService(executorServiceMock)
            .client(clientMock)
            .build()
        fetcherThread.start()
        fetcherThread.join()
        Assertions.assertThat(allmessages).isNotNull
        Assertions.assertThat(allmessages).hasSize(1)
        Assertions.assertThat(allmessages).contains("I am a message!")
        verify { queueMock.poll(1, TimeUnit.SECONDS) }
        verify(exactly = 1) { clientMock.connect() }
        verify(exactly = 1) { clientMock.stop() }
        verify { clientMock.isDone }
        verify { executorServiceMock.shutdownNow() }
    }

    @Test
    fun testRun_whenFetchMostlyTrackTimesamptAndCapacity_thenReturnOneMessage() {
        val allmessages: Set<String> = HashSet()
        val testMessage = "I am a message!"
        val testLimitTrack = "{\"limit\":{\"track\":26,\"timestamp_ms\":\"1579144966748\"}}"
        val stringStack = Stack<String>()
        stringStack.push(null)
        stringStack.push(testLimitTrack)
        stringStack.push(testLimitTrack)
        stringStack.push(testLimitTrack)
        stringStack.push(testLimitTrack)
        stringStack.push(testMessage)
        stringStack.push(testLimitTrack)
        stringStack.push(testLimitTrack)
        stringStack.push(testLimitTrack)
        every {
            queueMock.poll(1, TimeUnit.SECONDS)
        } returnsMany (stringStack)
        val fetcherThread = FetcherThread.builder()
            .stringLinkedBlockingQueue(queueMock)
            .capacity(1)
            .allMessages(allmessages)
            .executorService(executorServiceMock)
            .client(clientMock)
            .build()
        fetcherThread.start()
        fetcherThread.join()
        Assertions.assertThat(allmessages).isNotNull
        Assertions.assertThat(allmessages).hasSize(1)
        Assertions.assertThat(allmessages).contains("I am a message!")
        verify { queueMock.poll(1, TimeUnit.SECONDS) }
        verify(exactly = 1) { clientMock.connect() }
        verify(exactly = 1) { clientMock.stop() }
        verify { clientMock.isDone }
        verify { executorServiceMock.shutdownNow() }
    }

    @Test
    fun testRun_whenClientIsDone_thenReturnNoMessage() {
        val allmessages: Set<String> = HashSet()
        val event: Event = mockk()
        every { clientMock.isDone } returns true
        every { clientMock.exitEvent } returns event
        val fetcherThread = FetcherThread.builder()
            .stringLinkedBlockingQueue(queueMock)
            .capacity(1)
            .allMessages(allmessages)
            .executorService(executorServiceMock)
            .client(clientMock)
            .build()
        fetcherThread.start()
        fetcherThread.join()
        allmessages.shouldNotBeNull()
        allmessages.shouldBeEmpty()
        verify(exactly = 0) { queueMock.poll(1, TimeUnit.SECONDS) }
        verify { clientMock.connect() }
        verify { clientMock.isDone }
        verify { clientMock.exitEvent }
    }
}